### ISSUE TITLE
consul VM tests are failing with lost consensus

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#90613]
The [Consul VM test](https://github.com/NixOS/nixpkgs/blob/823d171956312ba2bdba0359f0448cd6d838cb75/nixos/tests/consul.nix#L101-L141) I wrote is failing  on `unstable` ([example](https://github.com/NixOS/nixpkgs/pull/90331/checks?check_run_id=769985233)):

```
client2 # [ 3552.015709] consul[767]:     [...] error="rpc error making call: No cluster leader"
building of '/nix/store/...-vm-test-run-consul.drv' timed out after 3600 seconds
```

The problem is that when machines reboot in quick succession, consensus (and cluster leadership) is lost permanently, and 2-of-3 servers cannot continue to serve requests.

I have debugged it and found that it is already broken since #84402 (`consul: 1.6.2 -> 1.7.2`), but it was noticed only now.

:red_circle: **This is why we should run the relevant tests for every PR and should show big red warning signs when they fail!**

In any case, I have bisected the problem to be commit https://github.com/hashicorp/consul/commit/c47dbffe1c5464d8aa9794e58687d894f1e50df9, introduced between Consul `1.7.0-beta1` and `1.7.0-beta2`.

I have filed an upstream bug: https://github.com/hashicorp/consul/issues/8118

I also found a way to make the test pass:

```diff
--- a/nixos/tests/consul.nix
+++ b/nixos/tests/consul.nix
@@ -129,7 +130,10 @@ in {
 
         # Wait for recovery.
         for m in machines:
-            m.wait_until_succeeds("[ $(consul members | grep -o alive | wc -l) == 5 ]")
+            m.wait_until_succeeds(
+                "[ $(consul operator raft list-peers | grep true | wc -l) == 3 ]"
+            )
```

But I don't think that's good, because I think this is a fundamental issue with Consul that affects operations (the 2-of-3 HA promise being broken), so I think it should be fundamentally fixed, and that the test is right.

NixOS 20.03 still has Consul `1.6` so it does not suffer from this bug, but the next NixOS release will.

---

CC @Mic92 @numkem @endocrimes @danbst @vdemeester @pradeepchhetri as previous consule package contributors.

**Comment by danbst** [comment]
IMO we shouldn't even use `consul` attr, but always refer to concrete version, like `consul_0_9` or `consul_1_6`. Seamless major consul upgrade is fine in docs, but disaster in reality (happened twice to me).

That doesn't quite solve this particular problem, I know.

**Comment by nh2** [comment]
> IMO we shouldn't even use `consul` attr, but always refer to concrete version, like `consul_0_9` or `consul_1_6`. Seamless major consul upgrade is fine in docs, but disaster in reality (happened twice to me).

@danbst Yes, the key thing is that Consul doesn't permit downgrades of its state in `/var`. I would thus also always recommend to upgrade non-point-releases separately from NixOS upgrades.

**Comment by endocrimes** [comment]
Honestly I kinda wanna make a few changes to most of the Hashi packages and tests soon to provide more stable versioning (and reliable testing). I'll start cleaning it up this weekend hopefully

**Comment by nh2** [comment]
Upstream has a great explanation and recommended fix: https://github.com/hashicorp/consul/issues/8118#issuecomment-645330040

Test fix and more added tests in PR #90701.

Key change: `autopilot.min_quorum = 3`.

**I recommend every Consul user to read that linked comment and the updated test** to understand the upstream-recommended rolling reboot/upgrade procedure.
